compiler_opt/rl/corpus.py

```diff
@@ -0,0 +1,180 @@
+  name: str
+
+  def cmd(self, **kwargs) -> List[str]:
+    """Retrieves the compiler execution options,
+    optionally adding configurable, pre-set options.
+    """
+    ret_cmd: List[str] = list(self._exec_cmd)
+    for k, substitutions in kwargs.items():
+      if k in self._xopts:
+        if substitutions is not None:
+          ret_cmd += [opt.format(**substitutions) for opt in self._xopts[k]]
+      else:
+        logging.fatal(
+            'Command line addition of \'%s\' requested '
+            'but no such extra option exists.', k)
+        raise ValueError
+    return ret_cmd
+
+  @classmethod
+  def _get(cls, data_path: str, additional_flags: Tuple[str, ...],
+           delete_flags: Tuple[str, ...], xopts: Dict[str, Tuple[str, ...]]):
+    module_paths: List[str] = _load_module_paths(
+        data_path, os.path.join(data_path, 'module_paths'))
+
+    is_thinlto: bool = _has_thinlto_index(module_paths)
+    has_cmd: bool = _has_cmd(module_paths)
+
+    # TODO: (b/233935329) Per-corpus *fdo profile paths can be read into
+    # {additional|delete}_flags here
+    meta = _load_metadata(os.path.join(data_path, 'metadata.json'))
+
+    xopts.update({'output': ('-o', '{path:s}')})
+
+    module_specs: List[ModuleSpec] = []
+
+    # Future: cmd_override could have per-module overrides if needed
+    if 'global_command_override' in meta:
+      cmd_override = tuple(meta['global_command_override'])
+    else:
+      cmd_override = None
+
+    # This takes ~7s for 30k modules
+    for module_path in module_paths:
+      exec_cmd = _load_and_parse_command(
+          cmd_file=(module_path + '.cmd') if has_cmd else None,
+          ir_file=module_path + '.bc',
+          thinlto_file=(module_path + '.thinlto.bc') if is_thinlto else None,
+          additional_flags=additional_flags,
+          delete_flags=delete_flags,
+          cmd_override=cmd_override)
+      module_specs.append(cls(tuple(exec_cmd), xopts, module_path))
+
+    return module_specs
+
+  @classmethod
+  @abc.abstractmethod
+  def get(cls, data_path: str, additional_flags: Tuple[str, ...],
+          delete_flags: Tuple[str, ...]) -> List[ModuleSpec]:
+    """Fetch a list of ModuleSpecs for the corpus at data_path
+
+    Args:
+      data_path: base directory of corpus
+      additional_flags: tuple of clang flags to add.
+      delete_flags: tuple of clang flags to remove.
+    """
+    raise NotImplementedError
+
+
+def _has_thinlto_index(module_paths: Iterable[str]) -> bool:
+  return tf.io.gfile.exists(next(iter(module_paths)) + '.thinlto.bc')
+
+
+def _has_cmd(module_paths: Iterable[str]) -> bool:
+  return tf.io.gfile.exists(next(iter(module_paths)) + '.cmd')
+
+
+def _load_module_paths(data_path, module_paths_path: str) -> List[str]:
+  with open(module_paths_path, 'r', encoding='utf-8') as f:
+    ret = [os.path.join(data_path, name.rstrip('\n')) for name in f]
+    if len(ret) == 0:
+      logging.fatal('%s is empty.', module_paths_path)
+      raise ValueError
```

Comment:
logging.fatal kills everything so don't need to raise an error here

Reply:
so either doing logging.fatal or raise ValueError with proper error message, recommend to raise ValueErrors to be consistent with the codebase

Reply:
ah ok, I put both since logging.fatal didn't seem kill the tests